### ISSUE TITLE
Try switching to Python alpine for the base image

### ISSUE THREAD

**Issue by marco-c** [mozilla/bugbug#292]
This might speed up the docker build task.

**Comment by trampfox** [comment]
Hi @marco-c, I started working on it. I'll shout if I'll need any kind of help :)

**Comment by trampfox** [comment]
@marco-c I’m struggling with the installation of `rs_parsepatch` on Alpine Linux. 
I managed to replace all the `apt-get` commands with the `apk` ones, but I got this error during the docker build: 
```
ERROR: Could not find a version that satisfies the requirement rs_parsepatch (from -r /requirements.txt (line 2)) (from versions: none)ERROR: No matching distribution found for rs_parsepatch (from -r /requirements.txt (line 2))
```

It seems that pip (inside the container) is not able to find `rs_parsepatch`, but this error doesn’t occur if I try to install it locally on my machine or with the `python:3.7-slim` as base image. 
I tried with Alpine linux version 3.9 and 3.10. 
Any idea?

**Comment by Lothiraldan** [comment]
@trampfox `rs_parsepatch` is packaged as pre-compiled wheels only (https://pypi.org/project/rs_parsepatch/#files), I don't know enough about Alpine to know how it affects the installation of pre-compiled wheels but you can pass the `--verbose` flag to debug the installation, it will explains why pip choose to not use the file ` rs_parsepatch-0.2.2-cp37-cp37m-manylinux1_x86_64.whl`.

**Comment by kanurag94** [comment]
I have done changes as per alpine requirements and they work fine on running them line by line on python-alpine.
But as soon as I try to compose docker for bugbug, service fails to build from: https://github.com/kanurag94/bugbug/blob/master/infra/dockerfile.commit_retrieval#L13

@Lothiraldan , Can you please help me figure out what's wrong here?

**Comment by Lothiraldan** [comment]
@kanurag94 Which error did you encountered?

If you encountered `/bin/sh: 1: apk: not found`, it's because docker-compose used the Debian-based bugbug-base image to build bugbug-commit-retrieval, try running the following command to solves it:

```bash
docker rmi -f mozilla/bugbug-base:latest
docker-compose build bugbug-base
docker-compose build bugbug-commit-retrieval
```

If you encountered the error:

```
Cloning into '/depot_tools'...
HEAD is now at 73065b20 depot_tools: Make some changes to make metrics collection compatible with Python 3.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   630    0   630    0     0   1143      0 --:--:-- --:--:-- --:--:--  1141
  0 2966k    0 16950    0     0  14713      0  0:03:26  0:00:01  0:03:25 14713rust-code-analysis
100 2966k  100 2966k    0     0   848k      0  0:00:03  0:00:03 --:--:-- 1258k
```

Try splitting the command into several `CMD` line to debug it.

**Comment by kanurag94** [comment]
@Lothiraldan Thanks, there was a small extra f in rm command. I'll open a PR since this works now locally

**Comment by Apoorv01017** [comment]
hello, can I work on this issue?

**Comment by marco-c** [comment]
Read CONTRIBUTING.md: you can work on any issue which has no open PR linked to it, so you can work on this one.

**Comment by Apoorv01017** [comment]
Hello, I have learned the basics of Docker. Can you please guide me on what to do next.

**Comment by marco-c** [comment]
Now that you know how to use Docker, the issue should be clear :)
You need to switch infra/dockerfile.base to use the Python Alpine image.

**Comment by Apoorv01017** [comment]
Can you please help me 
I am getting an error msg 
"The command '/bin/sh -c apt-get update &&                 apt-get install -y --no-install-recommends gcc g++ libgomp1 &&                 pip install --disable-pip-version-check --no-cache-dir -r /requirements.txt &&                 apt-get remove -y gcc g++ &&                 apt-get autoremove -y &&                 rm -rf /var/lib/apt/lists/*' returned a non-zero code: 127
"

**Comment by marco-c** [comment]
There's not enough info for me to be able to help you.

**Comment by Apoorv01017** [comment]
i just changed the first line of the file to FROM jfloff/alpine-python

**Comment by marco-c** [comment]
So the error is pretty obvious, on Alpine you can't run apt-get...

**Comment by Apoorv01017** [comment]
sorry for asking so many times
but i am having a hard time figuring this out
"apk add libgomp1 && \  " this statement is giving an error "libgomp1 (missing):
ERROR: unsatisfiable constraints:"

**Comment by kanurag94** [comment]
> sorry for asking so many times
> but i am having a hard time figuring this out
> "apk add libgomp1 && \ " this statement is giving an error "libgomp1 (missing):
> ERROR: unsatisfiable constraints:"

@Apoorv01017 use `libgomp`

**Comment by dotslash21** [comment]
> @marco-c I’m struggling with the installation of `rs_parsepatch` on Alpine Linux.
> I managed to replace all the `apt-get` commands with the `apk` ones, but I got this error during the docker build:
> 
> ```
> ERROR: Could not find a version that satisfies the requirement rs_parsepatch (from -r /requirements.txt (line 2)) (from versions: none)ERROR: No matching distribution found for rs_parsepatch (from -r /requirements.txt (line 2))
> ```
> 
> It seems that pip (inside the container) is not able to find `rs_parsepatch`, but this error doesn’t occur if I try to install it locally on my machine or with the `python:3.7-slim` as base image.
> I tried with Alpine linux version 3.9 and 3.10.
> Any idea?
> 
> https://github.com/trampfox/bugbug/tree/switch-to-alpine

I'm also facing the same, after some digging I came across this article [here](https://pythonspeed.com/articles/base-image-python-docker-images/).

TL;DR

Alpine uses a different C library, musl, instead of the more common glibc. So I think the binary wheels which are not compiled for Alpine wont work. So when installing `rs_parsepatch` using `--verbose` flag you will come across this `ERROR: rs_parsepatch-0.2.1-cp37-cp37m-manylinux1_x86_64.whl is not a supported wheel on this platform.`

**Comment by calixteman** [comment]
It seems to be ok with:
```docker
FROM alpine:latest

RUN apk add --update \
    python3 \
    musl \
    libgcc \
    gcompat

RUN echo 'manylinux1_compatible = True' > /usr/lib/python3.7/_manylinux.py
RUN python3 -m ensurepip && pip3 install --upgrade pip && pip3 install rs_parsepatch 
```

**Comment by dotslash21** [comment]
@marco-c, I've successfully converted the `dockerfile.base` and `dockerfile.commit_retrieval` files to use python 3.7 alpine image. But I'm stuck on `dockerfile.base-nlp` which gives this error when installing pyemd:
```
Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /usr/local/lib/python3.7/site-packages/numpy/core/_multiarray_umath.cpython-37m-x86_64-linux-gnu.so)
```
### dockerfile.base
```
FROM python:3.7-alpine

# Setup dependencies in a cacheable step
COPY requirements.txt /requirements.txt
RUN apk update && \
    apk add --no-cache gcc g++ libgomp libffi-dev && \
    echo 'manylinux1_compatible = True' > /usr/local/lib/python3.7/_manylinux.py && \
	pip install --disable-pip-version-check --no-cache-dir -r /requirements.txt && \
	apk del gcc g++ libgomp libffi-dev

COPY . /tmp/bugbug
RUN pip install --disable-pip-version-check --no-cache-dir /tmp/bugbug
```

### dockerfile.base-nlp
```
FROM base:latest

# Setup dependencies in a cacheable step
ADD extra-nlp-requirements.txt /

RUN apk update && \
    apk add --no-cache gcc g++ libgomp && \
    pip install --disable-pip-version-check --no-cache-dir --verbose -r /extra-nlp-requirements.txt && \
    apk del gcc g++ libgomp

RUN python -m spacy download en_core_web_sm
```

### Partial error log
```

Cleaning up...
  Removing source in /tmp/pip-install-zwu2dp0y/nltk
  Removing source in /tmp/pip-install-zwu2dp0y/pyemd
Removed pyemd==0.5.1 from https://files.pythonhosted.org/packages/c0/c5/7fea8e7a71cd026b30ed3c40e4c5ea13a173e28f8855da17e25271e8f545/pyemd-0.5.1.tar.gz#sha256=fc81c2116f8573e559dfbb8d73e0
3d9f73c22d0770559f406516984302e07e70 (from -r /extra-nlp-requirements.txt (line 3)) from build tracker '/tmp/pip-req-tracker-erfj50pq'
Removed build tracker '/tmp/pip-req-tracker-erfj50pq'
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
Exception information:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 153, in _main
    status = self.run(options, args)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 382, in run
    resolver.resolve(requirement_set)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py", line 201, in resolve
    self._resolve_one(requirement_set, req)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py", line 365, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py", line 313, in _get_abstract_dist_for
    req, self.session, self.finder, self.require_hashes
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/operations/prepare.py", line 224, in prepare_linked_requirement
    req, self.req_tracker, finder, self.build_isolation,
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/operations/prepare.py", line 49, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/distributions/source/legacy.py", line 39, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/req/req_install.py", line 563, in prepare_metadata
    self.metadata_directory = metadata_generator(self)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/operations/generate_metadata.py", line 124, in _generate_metadata_legacy
    command_desc='python setup.py egg_info',
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/utils/subprocess.py", line 242, in call_subprocess
    raise InstallationError(exc_msg)
pip._internal.exceptions.InstallationError: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
The command '/bin/sh -c apk update &&     apk add --no-cache gcc g++ libgomp &&     pip install --disable-pip-version-check --no-cache-dir --verbose -r /extra-nlp-requirements.txt &&     
apk del gcc g++ libgomp' returned a non-zero code: 1
```
I did some searching about the ld-linux-x86-64.so.2 file and came across `gcompat` package but building the image again after installing `gcompat` I'm greeted with another error:
```
Error relocating /usr/local/lib/python3.7/site-packages/numpy/core/_multiarray_umath.cpython-37m-x86_64-linux-gnu.so: backtrace: symbol not found
```

**Comment by dotslash21** [comment]
@marco-c it seems that switching to alpine based docker image is not straightforward and needs some essential pip packages to be rebuilt to run with MUSL. I can theoretically make the switch with successful build but the build time takes a huge hit, for instance numpy on the base image needs to be built from source for pyemd to work on the base-nlp image ([error report](https://github.com/mozilla/bugbug/issues/292#issuecomment-544253756)). There is a workaround though by using [glibc compatibility layer package](https://github.com/sgerrand/alpine-pkg-glibc). So should I use the glibc workaround?

**Comment by marco-c** [comment]
What is the size of the resulting image?
If the difference is big, it might be worth the complexity and longer build time. If the difference is not so big, maybe it's better not to do this at all.

**Comment by dotslash21** [comment]
@marco-c, here are the size of the respective images,
```
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
commit_retrieval    latest              05b65968d15e        5 seconds ago       906MB
base-nlp            latest              58bc95a3ff8f        29 minutes ago      1.18GB
base                latest              9f3725db74b8        54 minutes ago      807MB
python              3.7-alpine          b11d2a09763f        4 days ago          98.8MB
```

**Comment by marco-c** [comment]
OK, so the size difference is not that big, I think a faster build time is preferable in this case.

**Comment by jtlz2** [comment]
@dotslash21 Did you manage to find a way round the backtrace?

**Comment by dotslash21** [comment]
@jtlz2 Yes I did, see [this](https://github.com/mozilla/bugbug/issues/292#issuecomment-545923161).
